src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
         }
 
 #nullable enable
-        private void RemoveCallingConventionMismatches<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results, Cci.CallingConvention expectedConvention) where TMember : Symbol
+        private void RemoveCallingConventionMismatches<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results, in CallingConventionInfo expectedConvention) where TMember : Symbol
```

Comment:
>RemoveCallingConventionMismatches [](start = 21, length = 33)

Where rules to detect the mismatches are specified?

Reply:
[https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute)

---
In reply to: [488979189](https://github.com/dotnet/roslyn/pull/47651#discussion_r488979189) [](ancestors = 488979189)

Reply:
Perhaps I am missing something, but It look like that section only describes how to determine calling convention from the attribute, but doesn't specify what target calling convention can be. 

---
In reply to: [490568795](https://github.com/dotnet/roslyn/pull/47651#discussion_r490568795) [](ancestors = 490568795,488979189)

Reply:
> doesn't specify what the target calling convention can be.

I'm not certain I understand what you mean by this. Do you mean the format for a `CallingConventionInfo`? If so, that's a standard function pointer convention, CallKind + UnmanagedCallingConventionTypes if the CallKind is Unmanaged.

---
In reply to: [490580366](https://github.com/dotnet/roslyn/pull/47651#discussion_r490580366) [](ancestors = 490580366,490568795,488979189)

Reply:
> I'm not certain I understand what you mean by this. 

Let me rephrase. This function filters out candidates based on ```expectedConvention```. Where are the rules specified for detecting which candidate is a valid candidate and which candidate is an invalid candidate for the given ```expectedConvention``` (whatever information it contains)?

I would expect this function to:
- calculate CallingConventionInfo for a candidate. For a candidate with UnmanagedCallingConvention attribute, content of the attribute probably will contribute to this value, let's call it ```candidateConvention```  
- match the ```candidateConvention``` from the previous step against  ```expectedConvention``` by following some rules. Candidates that failed the match will be filtered out.

The question is where the rules are specified?
Another observation is that the code is structured in a way that it appears that candidates with UnmanagedCallingConvention attribute and candidates without it are not treated uniformly, as if rules were different. If we could restructure the code to unify the handling, I think we would be able to simplify the logic to some degree. However, these are all guesses at this point, because I do not know what rules exactly we are trying to follow here. 


---
In reply to: [490606880](https://github.com/dotnet/roslyn/pull/47651#discussion_r490606880) [](ancestors = 490606880,490580366,490568795,488979189)

Reply:
[https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#function-pointer-conversions](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#function-pointer-conversions)

We currently define the matching as "The calling convention of F0 is the same as the calling convention of F1." This could certainly be more precise (and I'll submit a PR to use the following rules in the definition):
* The CallKind must be the same
* If the CallKind is Unmanaged, the set of unmanaged convention types must be identical, ignoring order and duplicates.

I see your point about restructuring the checks, I'll do that.

---
In reply to: [490968551](https://github.com/dotnet/roslyn/pull/47651#discussion_r490968551) [](ancestors = 490968551,490606880,490580366,490568795,488979189)

Reply:
I guess the quote from [https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#allow-address-of-to-target-methods](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#allow-address-of-to-target-methods) covers the rules: "The calling convention of M is the same as the calling convention of F. This includes both the calling convention bit, as well as any calling convention flags specified in the unmanaged identifier."

---
In reply to: [491076666](https://github.com/dotnet/roslyn/pull/47651#discussion_r491076666) [](ancestors = 491076666,490968551,490606880,490580366,490568795,488979189)